crates/fuel-core/src/graphql_api/worker_service.rs

```diff
@@ -181,16 +204,116 @@ where
     }
 }
 
+trait DatabaseItemWithAmount {
+    type Storage: Mappable;
+
+    fn key(&self) -> <Self::Storage as Mappable>::Key;
+    fn amount(&self) -> Amount;
+}
+
+impl DatabaseItemWithAmount for Coin {
+    type Storage = CoinBalances;
+
+    fn key(&self) -> <Self::Storage as Mappable>::Key {
+        BalancesKey::new(&self.owner, &self.asset_id)
+    }
+
+    fn amount(&self) -> Amount {
+        self.amount
+    }
+}
+
+impl DatabaseItemWithAmount for Message {
+    type Storage = MessageBalances;
+
+    fn key(&self) -> <Self::Storage as Mappable>::Key {
+        *self.recipient()
+    }
+
+    fn amount(&self) -> Amount {
+        Self::amount(self)
+    }
+}
+
+trait BalanceIndexationUpdater: DatabaseItemWithAmount {
+    fn update_balances<T, F>(&self, tx: &mut T, updater: F) -> StorageResult<()>
+    where
+        <Self::Storage as Mappable>::Key: Sized + std::fmt::Display,
+        <Self::Storage as Mappable>::Value: Sized + std::fmt::Display,
+        <Self::Storage as Mappable>::OwnedValue: Default + std::fmt::Display,
+        T: OffChainDatabaseTransaction + StorageMutate<Self::Storage>,
+        F: Fn(
+            Cow<<Self::Storage as Mappable>::OwnedValue>,
+            Amount,
+        ) -> <Self::Storage as Mappable>::Value,
+        fuel_core_storage::Error: From<<T as StorageInspect<Self::Storage>>::Error>,
+    {
+        let key = self.key();
+        let amount = self.amount();
+        let storage = tx.storage::<Self::Storage>();
+        let current_balance = storage.get(&key)?.unwrap_or_default();
+        let prev_balance = current_balance.clone();
+        let new_balance = updater(current_balance, amount);
+
+        debug!(
+            %key,
+            %amount,
+            %prev_balance,
+            %new_balance,
+            "changing balance");
+
+        let storage = tx.storage::<Self::Storage>();
+        Ok(storage.insert(&key, &new_balance)?)
+    }
+}
+
+impl BalanceIndexationUpdater for Coin {}
+impl BalanceIndexationUpdater for Message {}
+
+fn process_balances_update<T>(
+    event: &Event,
+    block_st_transaction: &mut T,
+    balances_enabled: bool,
+) -> StorageResult<()>
+where
+    T: OffChainDatabaseTransaction,
+{
+    if !balances_enabled {
+        return Ok(());
+    }
+    match event {
+        Event::MessageImported(message) => message
+            .update_balances(block_st_transaction, |balance: Cow<u64>, amount| {
+                balance.saturating_add(amount)
```

Comment:
Good catch. Hmm, it can happen if anyone decide to create their own token with huge balances. I think the `Amount` should be `u128`.

Reply:
After I implemented this some of the integration tests started to fail, because they used some arbitrary transactions with arbitrary amounts and wallets.

I'll be fixing them to use the wallets available in the test environment. One such test is fixed [here](https://github.com/FuelLabs/fuel-core/pull/2383/commits/895d9da0928d3c799ae292d50be1853059cfc359). @MitchTurner, can you have a look at this and see if this fix makes sense? If so, I'll apply this to other tests.

Reply:
It turned out that the required changes to tests are too far reaching. I reverted the commits with modified test (474e207aa82c76e321862faa7a642b6fb824c1e6 de11071b8dbafb8b6d40048eaf847f5adc28078f 895d9da0928d3c799ae292d50be1853059cfc359) and added a note here (9848f64c844942274f2657c57dbb594c092ab449).

With balances type changed to `u128` we are a bit more safe, but still this is an issue worth discussing.

Reply:
I [changed](https://github.com/FuelLabs/fuel-core/pull/2383/files#diff-7e01ba9f6519c54fdacf70644783ac9d34a5739ef6d73e48d8eb594ceec34804) the asset balance to be `u128`, leaving the contract balance as is.

Reply:
Follow-up issue created: https://github.com/FuelLabs/fuel-core/issues/2428